compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -913,7 +913,59 @@ OMR::Power::TreeEvaluator::mstoreiEvaluator(TR::Node *node, TR::CodeGenerator *c
 TR::Register*
 OMR::Power::TreeEvaluator::mTrueCountEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    {
-   return TR::TreeEvaluator::unImpOpEvaluator(node, cg);
+   TR::Node *firstChild = node->getFirstChild();
+
+   TR_ASSERT_FATAL_WITH_NODE(node, firstChild->getDataType().getVectorLength() == TR::VectorLength128,
+                   "Only 128-bit vectors are supported %s", node->getDataType().toString());
+
+   TR::Register *srcReg = cg->evaluate(firstChild);
+   TR::Register *resReg = cg->allocateRegister(TR_GPR);
+
+   TR::Register *temp1 = cg->allocateRegister(TR_VRF);
+   TR::Register *temp2 = cg->allocateRegister(TR_VRF);
+
+   node->setRegister(resReg);
+
+   //pick shift distance based on vector element type
+   TR::DataType type = firstChild->getDataType().getVectorElementType();
+   int shift;
+
+   switch(type)
+     {
+     case TR::Int8:
+        shift = 3;
+	     break;
+     case TR::Int16:
+        shift = 4;
+	     break;
+     case TR::Int32:
+        shift = 5;
+        break;
+     case TR::Int64:
+        shift = 6;
+	     break;
+     default:
+	     TR_ASSERT_FATAL(false, "Unsupported vector type %s for mTrueCount\n", firstChild->getDataType().toString()); return NULL;
+     }
+
+   //get population counts of each half of input vector separately
+   generateTrg1Src1Instruction(cg, OMR::InstOpCode::vpopcntd, node, temp1, srcReg);
+   generateTrg1Src2ImmInstruction(cg, OMR::InstOpCode::xxpermdi, node, temp2, srcReg, srcReg, 2);
```

Comment:
what is this doubleWord-swapping for? temp2 is over-written in next instr anyway.